File: src/isre/graph/builder.ts
import { IntentGraph, IntentNode, SemanticRelation } from '../../types';
import { SemanticPrimitive, IntentType, EdgeType } from '../types';

export class IntentGraphBuilder {
  /**
   * Transforms SemanticPrimitives into a structured IntentGraph.
   * Implements explicit conflict detection and representation.
   */
  async buildFromPrimitives(primitives: SemanticPrimitive[]): Promise<IntentGraph> {
    const nodes: IntentNode[] = [];
    const edges: SemanticRelation[] = [];

    // 1. Generate Nodes
    for (const prim of primitives) {
      const nodeType = this.inferIntentType(prim);
      const node: IntentNode = {
        id: `node_${prim.id}`,
        label: prim.concept,
        confidence: prim.semanticWeight,
        attributes: {
          type: nodeType,
          semanticPayload: [prim],
          conflictMarkers: []
        }
      };
      nodes.push(node);
    }

    // 2. Establish basic causal/temporal edges (sequential by default)
    for (let i = 0; i < nodes.length - 1; i++) {
      edges.push({
        sourceId: nodes[i].id,
        targetId: nodes[i + 1].id,
        relationType: EdgeType.TEMPORAL,
        weight: 1.0
      });
    }

    // 3. Detect and represent conflicts
    this.detectConflicts(nodes);

    return {
      nodes,
      edges,
      rootIntent: nodes.length > 0 ? nodes[0].id : '',
      confidenceScore: this.calculateGraphConfidence(nodes)
    };
  }

  private inferIntentType(primitive: SemanticPrimitive): IntentType {
    const concept = primitive.concept.toLowerCase();
    if (concept.includes("action") || concept.includes("goal")) return IntentType.GOAL;
    if (concept.includes("query") || concept.includes("?")) return IntentType.QUERY;
    if (concept.includes("constraint") || concept.includes("must") || concept.includes("only")) return IntentType.CONSTRAINT;
    if (concept.includes("emotion")) return IntentType.EMOTION;
    return IntentType.CONTEXT;
  }

  private detectConflicts(nodes: IntentNode[]) {
    for (let i = 0; i < nodes.length; i++) {
      for (let j = i + 1; j < nodes.length; j++) {
        const n1 = nodes[i];
        const n2 = nodes[j];

        if (this.areConflicting(n1, n2)) {
          const conflictInfo = {
            type: "semantic_opposition",
            partnerId: n2.id,
            description: `Conflict between ${n1.id} and ${n2.id}`
          };
          n1.attributes.conflictMarkers.push(conflictInfo);
          
          n2.attributes.conflictMarkers.push({
            type: "semantic_opposition",
            partnerId: n1.id,
            description: `Conflict between ${n2.id} and ${n1.id}`
          });
        }
      }
    }
  }

  private areConflicting(n1: IntentNode, n2: IntentNode): boolean {
    // Example logic: "fast" vs "slow"
    // We assume payload is available in attributes
    const concepts1 = (n1.attributes.semanticPayload as SemanticPrimitive[]).map(p => p.concept);
    const concepts2 = (n2.attributes.semanticPayload as SemanticPrimitive[]).map(p => p.concept);

    const opposites: Record<string, string> = {
      "attribute_fast": "attribute_slow",
      "action_move_fast": "action_move_slow"
    };

    // Check bidirectional
    for (const c1 of concepts1) {
      if (opposites[c1] && concepts2.includes(opposites[c1])) return true;
    }
    for (const c2 of concepts2) {
      if (opposites[c2] && concepts1.includes(opposites[c2])) return true;
    }

    return false;
  }

  private calculateGraphConfidence(nodes: IntentNode[]): number {
    if (nodes.length === 0) return 0;
    const totalConfidence = nodes.reduce((sum, node) => sum + node.confidence, 0);
    return totalConfidence / nodes.length;
  }
}

File: isre/graph/builder.py
from typing import List, Dict, Any, Optional
import uuid
from ..models.primitives import SemanticPrimitive
from ..models.intent import IntentNode, IntentEdge, IntentGraph
from ..types import IntentType, EdgeType

class IntentGraphBuilder:
    """
    Transforms SemanticPrimitives into a structured IntentGraph.
    Implements explicit conflict detection and representation.
    """

    def build_from_primitives(self, primitives: List[SemanticPrimitive]) -> IntentGraph:
        """
        Creates a basic intent graph where each primitive becomes a node.
        In a more advanced implementation, this would involve heuristic or LLM-guided clustering.
        """
        graph = IntentGraph()
        nodes = []

        # 1. Generate Nodes
        # For the prototype, we assume each meaningful concept warrants an intent node.
        # We categorize them based on concept prefixes or metadata if available.
        for prim in primitives:
            node_type = self._infer_intent_type(prim)
            node = IntentNode(
                id=f"node_{prim.id}",
                type=node_type,
                semantic_payload=[prim],
                activation_level=prim.semantic_weight
            )
            graph.add_node(node)
            nodes.append(node)

        # 2. Establish basic causal/temporal edges (sequential by default for now)
        for i in range(len(nodes) - 1):
            edge = IntentEdge(
                source_id=nodes[i].id,
                target_id=nodes[i+1].id,
                relationship_type=EdgeType.TEMPORAL,
                semantic_label="sequenced_intent"
            )
            graph.add_edge(edge)

        # 3. Detect and represent conflicts
        self._detect_conflicts(graph)

        return graph

    def _infer_intent_type(self, primitive: SemanticPrimitive) -> IntentType:
        """Heuristic to map semantic concepts to intent types."""
        concept = primitive.concept.lower()
        if "action" in concept or "goal" in concept:
            return IntentType.GOAL
        if "query" in concept or "?" in concept:
            return IntentType.QUERY
        if "constraint" in concept or "must" in concept or "only" in concept:
            return IntentType.CONSTRAINT
        if "emotion" in concept:
            return IntentType.EMOTION
        return IntentType.CONTEXT

    def _detect_conflicts(self, graph: IntentGraph):
        """
        Explicitly identifies and marks conflicts between nodes.
        Requirement 2.4: Explicitly represent conflicts.
        """
        node_list = list(graph.nodes.values())
        for i in range(len(node_list)):
            for j in range(i + 1, len(node_list)):
                n1 = node_list[i]
                n2 = node_list[j]
                
                # Simple conflict heuristic: opposite semantic concepts
                if self._are_conflicting(n1, n2):
                    conflict_info = {
                        "type": "semantic_opposition",
                        "partner_id": n2.id,
                        "description": f"Conflict between {n1.id} and {n2.id}"
                    }
                    n1.conflict_markers.append(conflict_info)
                    
                    # Mutual marking
                    n2.conflict_markers.append({
                        "type": "semantic_opposition",
                        "partner_id": n1.id,
                        "description": f"Conflict between {n2.id} and {n1.id}"
                    })

    def _are_conflicting(self, n1: IntentNode, n2: IntentNode) -> bool:
        """Determines if two nodes have conflicting semantics."""
        # Example logic: "fast" vs "slow" in the same context
        concepts1 = [p.concept for p in n1.semantic_payload]
        concepts2 = [p.concept for p in n2.semantic_payload]
        
        opposites = {
            "attribute_fast": "attribute_slow",
            "action_move_fast": "action_move_slow"
        }
        
        for c1 in concepts1:
            for c2 in concepts2:
                if opposites.get(c1) == c2 or opposites.get(c2) == c1:
                    return True
        return False
